### PR TITLE
Issue113

### PR DESCRIPTION
**Pull request by vixie**

### CODE REVIEW

File: dnsdbq.c [hunk 1]
```diff
@@ -350,11 +363,10 @@ main(int argc, char *argv[]) {
 	/* global dynamic initialization. */
 	ideal_buffer = 4 * (size_t) sysconf(_SC_PAGESIZE);
 	gettimeofday(&now, NULL);
-	program_name = strrchr(argv[0], '/');
-	if (program_name != NULL)
-		program_name++;
-	else
+	if ((program_name = strrchr(argv[0], '/')) == NULL)
```
Comment (djw1149) [inline dnsdbq.c]: why do an assignment inside an if statement?

File: dnsdbq.c [hunk 2]
```diff
@@ -819,20 +830,26 @@ main(int argc, char *argv[]) {
...
+			.bailiwick = bailiwick,
+			.pfxlen = pfxlen,
+			.after = after,
+			.before = before
+		};
 		server_setup();
 		make_curl();
-		pdns_query(command, after, before);
-		DESTROY(command);
+		pdns_query((query_ct)&q);
```
Comment (djw1149) [inline dnsdbq.c]: I don't think that the cast (query_ct) is needed

File: dnsdbq.c [hunk 3]
```diff
@@ -1246,14 +1268,35 @@ do_batch(FILE *f, u_long after, u_long before) {
...
 		} else {
-			io_engine(0);
+			/* manage batch-level defaults as -A and -B. */
+			if (q.after == 0)
+				q.after = after;
+			if (q.before == 0)
+				q.before = before;
+
+			/* start one or two curl jobs based on this search. */
+			query_launcher((query_ct)&q, writer);
```
Comment (djw1149) [inline dnsdbq.c]: I don't think that the cast (query_ct) is needed

File: dnsdbq.c [hunk 4]
```diff
@@ -334,27 +350,30 @@ static bool sort_bydata = false;
 static writer_t writers = NULL;
 static int exit_code = 0; /* hopeful */
 static size_t ideal_buffer;
+static bool iso = false;
```
Comment (djw1149) [inline dnsdbq.c]: I assume this is not a switch to use OSI networking, so maybe clarify the variable name or comment it?

File: dnsdbq.c [hunk 5]
```diff
@@ -295,10 +310,11 @@ static const struct verb verbs[] = {
 #define	MAX_KEYS 5
 #define	MAX_JOBS 8
 
-#define CREATE(p, s) if ((p) != NULL) { my_panic("non-NULL ptr"); }	\
-	else if (((p) = malloc(s)) == NULL) { my_panic("malloc failed"); } \
+#define CREATE(p, s) if ((p) != NULL) { my_panic(false, "non-NULL ptr"); } \
+	else if (((p) = malloc(s)) == NULL) { my_panic(true, "malloc"); } \
 	else { memset((p), 0, s); }
 #define DESTROY(p) { if ((p) != NULL) { free(p); (p) = NULL; } }
+#define DEBUG(ge, va) { if (debug_level >= (ge)) debug va; }
```
Comment (djw1149) [inline dnsdbq.c]: Isn't there a better "new" C way to define such a macro?
#define DEBUG(ge, ...) { if (debug_level >= (ge)) debug__VA_ARGS__; }
though if you change the callers to remove the inner paratheses then a clearer variation can be used:
#define DEBUG(ge, ...) { if (debug_level >= (ge)) debug(__VA_ARGS__); }

Tested with clang and gcc-9 on macOS

File: dnsdbq.c [hunk 6]
```diff
@@ -1283,11 +1333,90 @@ do_batch(FILE *f, u_long after, u_long before) {
 	}
 }
 
+/* batch_parse -- turn one line from a -f batch into a (struct query).
```
Comment (djw1149) [inline dnsdbq.c]: I would like a comment syntax (perhaps #) and for it to ignore blank lines

File: dnsdbq.c [hunk 7]
```diff
@@ -2202,21 +2309,28 @@ io_engine(int jobs) {
...
-		if (debuglev > 3)
-			fprintf(stderr, "...waiting (still %d)\n", still);
+		DEBUG(4, (true, "...waiting (still %d)\n", still));
 		numfds = 0;
 		if (curl_multi_wait(multi, NULL, 0, 0, &numfds) != CURLM_OK)
 			break;
 		if (numfds == 0) {
-			if (++repeats > 1)
-				usleep(100000);
+			if (++repeats > 1) {
```
Comment (djw1149) [inline dnsdbq.c]: Maybe add comment: /* Wait 100ms even if interrupted.  After an interrupt, resume waiting for remaining time. */

File: dnsdbq.c [hunk 8]
```diff
@@ -2785,27 +2866,21 @@ timecmp(u_long a, u_long b) {
 	return (0);
 }
 
-/* time_print -- format one (possibly relative) timestamp.
+/* time_str -- format one (possibly relative) timestamp (returns static string)
```
Comment (djw1149) [inline dnsdbq.c]: While the callers of this revised function look much cleaner, I don't like the risk of a caller not realizing it is static.  Is there a function naming convention to indicate *not* reentrant etc.?

**Comment by djw1149** [comment]
CIRCL appears broken:
$ dnsdbq -u circl -r fsi.io -dd
debug: thing = 'fsi.io'
debug: batching=0, merge=0
error: the CIRCL system only understands 'lookup'

try   dnsdbq -h   for a short description of program usage.
debug: about to call exit(1)

**Comment by vixie** [comment]
On Wednesday, 12 February 2020 14:18:21 UTC djw1149 wrote:
> ...
> +	if ((program_name = strrchr(argv[0], '/')) == NULL)
> 
> why do an assignment inside an if statement?

tersity is a virtue during one-time initializations. every assignment in C is 
also an expression.

-- 
Paul

**Comment by vixie** [comment]
On Wednesday, 12 February 2020 15:43:48 UTC djw1149 wrote:
> ...
> +static bool iso = false;
> 
> I assume this is not a switch to use OSI networking, so maybe clarify the
> variable name or comment it?

ok, will change it to iso8601.

-- 
Paul

**Comment by djw1149** [comment]
> On Wednesday, 12 February 2020 14:18:21 UTC djw1149 wrote: ... +	if ((program_name = strrchr(argv[0], '/')) == NULL) why do an assignment inside an if statement?
> tersity is a virtue during one-time initializations. every assignment in C is also an expression.
> […](#)

I am quite aware that assignment is an expression.  But assignment in an if statement has been frowned upon for decades.  Yes, having the extra parens turns off the typical warning...

**Comment by vixie** [comment]
On Wednesday, 12 February 2020 16:27:46 UTC djw1149 wrote:
> ...
> 
> I would like a comment syntax (perhaps #) and for it to ignore blank lines

i think that would be a mistake. -f and -ff are principally used to speak to 
dnsdbq over popen() from a parent process. i would not want to ignore anything 
we hear from the parent process; every line should be something we can answer 
with a -- or a ++/-- framed response.

-- 
Paul

**Comment by djw1149** [comment]
> On Wednesday, 12 February 2020 16:27:46 UTC djw1149 wrote: ... I would like a comment syntax (perhaps #) and for it to ignore blank lines
> i think that would be a mistake. -f and -ff are principally used to speak to dnsdbq over popen() from a parent process. i would not want to ignore anything we hear from the parent process; every line should be something we can answer with a -- or a ++/-- framed response.

What about a use case with a static file of queries to make?  Maybe the user wants to edit the input file to disable certain queries or comment what is happening.  The user could send the input file through a filter, I guess: `perl -ne 'print if !/^#/ and /./' | dnsdbq -f ...`

**Comment by vixie** [comment]
they can use sed.

⁣Get BlueMail for Android ​

**Comment by vixie** [comment]
On Wednesday, 12 February 2020 16:40:31 UTC djw1149 wrote:
> ...
>  		if (numfds == 0) {
> -			if (++repeats > 1)
> -				usleep(100000);
> +			if (++repeats > 1) {
> 
> Maybe add comment: /* Wait 100ms even if interrupted.  After an interrupt,
> resume waiting for remaining time. */

comment will be added. however, are you looking at the right source code? in 
the issue113 branch this block of code calls nanosleep() in a loop.

-- 
Paul

**Comment by vixie** [comment]
On Wednesday, 12 February 2020 16:47:22 UTC djw1149 wrote:
> ...
> +/* time_str -- format one (possibly relative) timestamp (returns static
> string)
> 
> While the callers of this revised function look much cleaner, I don't like
> the risk of a caller not realizing it is static.  Is there a function
> naming convention to indicate *not* reentrant etc.?

there is no such convention in my experience. i think the comment on the 
function itself is adequate, since so many other time related functions also 
return static values, and since anyone calling it can certainly know that the 
string has to be stored somewhere and they aren't being asked to free() it.

-- 
Paul

**Comment by djw1149** [comment]
> On Wednesday, 12 February 2020 16:40:31 UTC djw1149 wrote: ... if (numfds == 0) { -	if (++repeats > 1) -	usleep(100000); +	if (++repeats > 1) { Maybe add comment: /* Wait 100ms even if interrupted. After an interrupt, resume waiting for remaining time. */
> comment will be added. however, are you looking at the right source code? in the issue113 branch this block of code calls nanosleep() in a loop.

Yes.  I don't know why github shows my comment on the old code.  The reason for the comment is the "while nanosleep ... req = rem;" is not obvious if you don't know that nanosleep updates &rem with the remaining time.

**Comment by djw1149** [comment]
> there is no such convention in my experience. i think the comment on the function itself is adequate, since so many other time related functions also return static values, and since anyone calling it can certainly know that the string has to be stored somewhere and they aren't being asked to free() it.

Fine.

**Comment by vixie** [comment]
On Wednesday, 12 February 2020 14:25:57 UTC djw1149 wrote:
> djw1149 commented on this pull request.
> 
> >  		server_setup();
> 
>  		make_curl();
> -		pdns_query(command, after, before);
> -		DESTROY(command);
> +		pdns_query((query_ct)&q);
> 
> I don't think that the cast (query_ct) is needed

ok, i will remove this. (i had trouble with const vs. not, earlier, but it's 
gone now.)

-- 
Paul

**Comment by vixie** [comment]
On Wednesday, 12 February 2020 17:15:28 UTC djw1149 wrote:
> CIRCL appears broken:
> $ dnsdbq -u circl -r fsi.io -dd
> debug: thing = 'fsi.io'
> debug: batching=0, merge=0
> error: the CIRCL system only understands 'lookup'

will be fixed. it was this:

@@ -3300,7 +3302,7 @@ circl_status(reader_t reader __attribute__((unused))) {
static const char *
circl_validate_verb(const char *verb_name) {
/* Only "lookup" is valid */
-       if (strcasecmp(verb_name, "lookup") == 0)
+       if (strcasecmp(verb_name, "lookup") != 0)
return ("the CIRCL system only understands 'lookup'");
return (NULL);
}

-- 
Paul

**Comment by vixie** [comment]
On Wednesday, 12 February 2020 15:50:16 UTC djw1149 wrote:
> Isn't there a better "new" C way to define such a macro?
> #define DEBUG(ge, ...) { if (debug_level >= (ge)) debug__VA_ARGS__; }
> though if you change the callers to remove the inner paratheses then a
> clearer variation can be used: #define DEBUG(ge, ...) { if (debug_level >=
> (ge)) debug(__VA_ARGS__); }

i had no idea. i took your second suggestion above.

-- 
Paul

**Pull request merged**